File: src/CardanoSwaps/TwoWaySwap.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE StrictData #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE OverloadedStrings #-}

module CardanoSwaps.TwoWaySwap
  (
    -- * On-Chain Data Types
    SwapDatum(..)
  , SwapRedeemer(..)
  , BeaconRedeemer(..)
  , SwapAsset(..)

    -- * Contracts
  , swapScript
  , swapScriptSize
  , swapValidatorHash
  , beaconScript
  , beaconScriptSize
  , beaconCurrencySymbol

    -- * Beacon Names
  , genPairBeaconName
  , genAssetBeaconName

    -- * Datums
  , genSwapDatum

    -- * Helpers
  , getRequiredSwapDirection
  ) where

import qualified PlutusTx
import qualified PlutusTx.Prelude as PlutusTx
import qualified PlutusTx.Builtins as Builtins
import GHC.Generics (Generic)
import qualified Data.Map as Map
import Data.Aeson
import qualified Plutus.Script.Utils.Scripts as PV3
import qualified PlutusLedgerApi.V3 as PV3

import CardanoSwaps.Utils
import CardanoSwaps.Blueprints

-------------------------------------------------
-- On-Chain Data Types
-------------------------------------------------
data SwapDatum = SwapDatum
  { beaconId :: CurrencySymbol -- ^ `CurrencySymbol` for the `beaconScript`.
  , pairBeacon :: TokenName -- ^ The pair beacon's `TokenName` for this trading pair.
  , asset1Id :: CurrencySymbol -- ^ The `CurrencySymbol` for asset1.
  , asset1Name :: TokenName -- ^ The `TokenName` for asset1.
  , asset1Beacon :: TokenName -- ^ The asset beacon's `TokenName` for asset1.
  , asset2Id :: CurrencySymbol -- ^ The `CurrencySymbol` for asset2.
  , asset2Name :: TokenName -- ^ The `TokenName` for asset2.
  , asset2Beacon :: TokenName -- ^ The asset beacon's `TokenName` for asset2.
  , asset1Price :: PlutusRational -- ^ The price to take asset1 as a fraction (Asset2/Asset1).
  , asset2Price :: PlutusRational -- ^ The price to take asset2 as a fraction (Asset1/Asset2).
  , prevInput :: Maybe TxOutRef -- ^ The corresponding swap input's output reference.
  , expiration :: Maybe POSIXTime -- ^ The order's expiration.
  } deriving (Generic,Show,Eq)

instance ToJSON SwapDatum where
  toJSON SwapDatum{..} = 
    object [ "beacon_id" .= show beaconId
           , "pair_beacon" .= showTokenName pairBeacon
           , "asset1_id" .= show asset1Id
           , "asset1_name" .= showTokenName asset1Name
           , "asset1_beacon" .= showTokenName asset1Beacon
           , "asset2_id" .= show asset2Id
           , "asset2_name" .= showTokenName asset2Name
           , "asset2_beacon" .= showTokenName asset2Beacon
           , "asset1_price" .= asset1Price 
           , "asset2_price" .= asset2Price 
           , "prev_input" .= prevInput
           , "expiration" .= expiration
           ]

data SwapRedeemer
  -- | Spend the swap as the owner using the beacon script as a minting policy.
  = SpendWithMint 
  -- | Spend the swap as the owner using the beacon script as a staking validator. This is only
  -- for when no beacons need to be minted or burned in the transaction.
  | SpendWithStake 
  -- | Take asset1 and deposit asset2.
  | TakeAsset1 
  -- | Take asset2 and deposit asset1.
  | TakeAsset2 
  deriving (Generic,Show)

data BeaconRedeemer
  -- | Register the beacon script for staking execution. This can only be done once. The credential
  -- connat be de-registered or delegated.
  = RegisterBeaconScript
  -- | Execute the beacon script as a minting policy. Used anytime beacons must be minted or burned.
  | CreateOrCloseSwaps 
  -- | Execute the beacon script as a staking validtor. Used anytime beacons do not need to be
  -- minted or burned.
  | UpdateSwaps
  deriving (Generic,Show)

-- | The pre-image for beacon token names. Every beacon name is:
--
-- > sha2_256 ( serialiseData ( toBuiltinData swapAsset ) )
--
-- The Data encoding tags each constructor and length-prefixes each field, so the encoding is
-- injective: distinct assets/pairs can never produce the same pre-hash bytes, and beacon names
-- cannot collide without a sha2_256 collision.
--
-- The constructor order must match the on-chain SwapAsset type exactly since it determines the
-- constructor tag in the Data encoding: append only, never reorder or remove constructors. The
-- tags may coincide with the one-way swap's SwapAsset tags; that is harmless because the two
-- protocols use different beacon policies. This type never appears in a datum or redeemer; it is
-- internal to the beacon name derivation.
data SwapAsset
  -- | An asset in the pair. On-chain: @Asset@ (Constr tag 121).
  = Asset CurrencySymbol TokenName
  -- | The trading pair, already sorted: asset1 < asset2 lexicographically. On-chain:
  -- @SortedPair@ (Constr tag 122).
  | SortedPair CurrencySymbol TokenName CurrencySymbol TokenName
  deriving (Generic,Show,Eq)

PlutusTx.unstableMakeIsData ''SwapDatum
PlutusTx.unstableMakeIsData ''SwapRedeemer
PlutusTx.unstableMakeIsData ''BeaconRedeemer
PlutusTx.unstableMakeIsData ''SwapAsset

-------------------------------------------------
-- Contracts
-------------------------------------------------
swapScript :: SerialisedScript
swapScript = parseScriptFromCBOR $ blueprints Map.! "two_way_swap.swap_script"

swapScriptSize :: Integer
swapScriptSize = getScriptSize swapScript

swapValidatorHash :: PV3.ValidatorHash
swapValidatorHash = PV3.ValidatorHash $ PV3.getScriptHash $ scriptHash swapScript

beaconScript :: SerialisedScript
beaconScript =
  applyArguments
    (parseScriptFromCBOR $ blueprints Map.! "two_way_swap.beacon_script")
    [PlutusTx.toData swapValidatorHash]

beaconScriptSize :: Integer
beaconScriptSize = getScriptSize beaconScript

beaconCurrencySymbol :: PV3.CurrencySymbol
beaconCurrencySymbol = PV3.CurrencySymbol $ PV3.getScriptHash $ scriptHash beaconScript

-------------------------------------------------
-- Beacon Names
-------------------------------------------------
-- | Hash the Data encoding of a `SwapAsset` to create a beacon name.
hashSwapAsset :: SwapAsset -> TokenName
hashSwapAsset = TokenName . PlutusTx.sha2_256 . Builtins.serialiseData . PlutusTx.toBuiltinData

-- | Generate the beacon asset name by hashing the serialised `SortedPair`. The trading pair is
-- first sorted so that the beacon name is independent of the ordering. ADA (empty policy id and
-- asset name) needs no special handling since the Data encoding length-prefixes every field.
--
-- > sha2_256 ( serialiseData ( SortedPair asset1Id asset1Name asset2Id asset2Name ) )
genPairBeaconName :: AssetConfig -> AssetConfig -> TokenName
genPairBeaconName assetX assetY =
  let ((sym1,name1),(sym2,name2)) =
       if assetY < assetX then (assetY,assetX) else (assetX,assetY)
  in hashSwapAsset $ SortedPair sym1 name1 sym2 name2

-- | Generate the beacon asset name by hashing the serialised `Asset`.
--
-- > sha2_256 ( serialiseData ( Asset assetId assetName ) )
genAssetBeaconName :: AssetConfig -> TokenName
genAssetBeaconName (sym,name) = hashSwapAsset $ Asset sym name

-------------------------------------------------
-- Datums
-------------------------------------------------
-- | Create the datum for a swap. The format should be:
-- 
-- > genSwapDatum (firstAsset,secondAsset) firstPrice secondPrice mPrev
--
-- Which asset is first or second does not matter; just make sure the
-- first price corresponds to __taking__ the first asset and the second price
-- corresponds to __taking__ the second asset.
genSwapDatum 
  :: TwoWayPair 
  -> PlutusRational 
  -> PlutusRational 
  -> Maybe TxOutRef 
  -> Maybe POSIXTime
  -> SwapDatum
genSwapDatum (firstAsset,secondAsset) firstPrice secondPrice mPrev mExpir =
    let (asset1,asset2) = 
          if firstAsset < secondAsset 
          then (firstAsset,secondAsset) 
          else (secondAsset,firstAsset)
        (asset1Price,asset2Price) = 
          if asset1 == firstAsset 
          then (firstPrice,secondPrice)
          else (secondPrice,firstPrice)
    in SwapDatum 
        { beaconId = beaconCurrencySymbol
        , pairBeacon = genPairBeaconName asset1 asset2
        , asset1Id = fst asset1
        , asset1Name = snd asset1
        , asset1Beacon = genAssetBeaconName asset1
        , asset2Id = fst asset2
        , asset2Name = snd asset2
        , asset2Beacon = genAssetBeaconName asset2
        , asset1Price = asset1Price
        , asset2Price = asset2Price
        , prevInput = mPrev
        , expiration = mExpir
        }

-------------------------------------------------
-- Helpers
-------------------------------------------------
-- | Get the required two-way swap redeemer based on the desired swap direction.
getRequiredSwapDirection :: OfferAsset -> AskAsset -> SwapRedeemer
getRequiredSwapDirection (OfferAsset offer) (AskAsset ask)
  | offer == min offer ask = TakeAsset1
  | otherwise = TakeAsset2

File: scalus-tests/sdk/TwoWaySwap.scala
package cardanoswaps

import scalus.*
import scalus.uplc.builtin.ByteString
import scalus.uplc.builtin.Data.{ToData, toData}
import scalus.cardano.blueprint.Blueprint
import scalus.cardano.ledger.Script
import scalus.cardano.onchain.plutus.v3.TxOutRef
import scalus.cardano.onchain.plutus.prelude.{Rational as _, *}
import scalus.uplc.Program

import java.nio.file.{Files, Path}
import java.security.MessageDigest

// -- On-Chain Data Types --

case class TwoWaySwapDatum(
    beaconId: ByteString,
    pairBeacon: ByteString,
    asset1Id: ByteString,
    asset1Name: ByteString,
    asset1Beacon: ByteString,
    asset2Id: ByteString,
    asset2Name: ByteString,
    asset2Beacon: ByteString,
    asset1Price: Rational,
    asset2Price: Rational,
    prevInput: Option[TxOutRef],
    expiration: Option[BigInt]
) derives ToData

enum TwoWaySwapRedeemer derives ToData:
    case SpendWithMint, SpendWithStake, TakeAsset1, TakeAsset2

enum TwoWayBeaconRedeemer derives ToData:
    case RegisterBeaconScript, CreateOrCloseSwaps, UpdateSwaps

// -- SDK --

object TwoWaySwap:
    private val blueprintJson = Files.readString(java.nio.file.Path.of("aiken/plutus.json"))
    private val blueprint = Blueprint.fromJson(blueprintJson)

    // Load swap script
    private val rawSwapProgram =
        blueprint.validators
            .find(_.title == "two_way_swap.swap_script")
            .get
            .compiledCode
            .map(Program.fromCborHex)
            .get

    val swapScript = Script.PlutusV2(rawSwapProgram.cborByteString)
    val swapScriptHash = swapScript.scriptHash

    // Load beacon script (parameterized by swap validator hash)
    private val rawBeaconProgram =
        blueprint.validators
            .find(_.title == "two_way_swap.beacon_script")
            .get
            .compiledCode
            .map(Program.fromCborHex)
            .get

    private val appliedBeaconProgram = rawBeaconProgram $ swapScriptHash.toData
    val beaconScript = Script.PlutusV2(appliedBeaconProgram.cborByteString)
    val beaconCurrencySymbol = beaconScript.scriptHash

    // Unapplied beacon script (for hash verification against blueprint)
    val rawBeaconScript: Script = Script.PlutusV2(rawBeaconProgram.cborByteString)
    val rawBeaconScriptHash: ByteString = rawBeaconScript.scriptHash

    // -- Beacon Name Generation --

    private def sha256(data: Array[Byte]): ByteString =
        val digest = MessageDigest.getInstance("SHA-256")
        ByteString.fromArray(digest.digest(data))

    /** Compare two asset configs lexicographically (policyId first, then name).
      * Matches Haskell Ord for (CurrencySymbol, TokenName).
      */
    private def compareAssets(
        id1: ByteString, name1: ByteString,
        id2: ByteString, name2: ByteString
    ): Int =
        val c = id1.toHex.compareTo(id2.toHex)
        if c != 0 then c else name1.toHex.compareTo(name2.toHex)

    /** Sort the pair so beacon name is order-independent, then hash.
      * ADA's empty policy ID is replaced with 0x00 after sorting.
      */
    def genPairBeaconName(
        assetXId: ByteString, assetXName: ByteString,
        assetYId: ByteString, assetYName: ByteString
    ): ByteString =
        // Sort: if Y < X then swap
        val (s1Id, s1Name, s2Id, s2Name) =
            if compareAssets(assetYId, assetYName, assetXId, assetXName) < 0
            then (assetYId, assetYName, assetXId, assetXName)
            else (assetXId, assetXName, assetYId, assetYName)
        val sym1 = if s1Id.isEmpty then ByteString.fromHex("00") else s1Id
        val sym2 = if s2Id.isEmpty then ByteString.fromHex("00") else s2Id
        sha256((sym1 ++ s1Name ++ sym2 ++ s2Name).bytes)

    /** Hash assetId ++ assetName (no prefix). */
    def genAssetBeaconName(assetId: ByteString, assetName: ByteString): ByteString =
        sha256((assetId ++ assetName).bytes)

    // -- Datum Construction --

    /** Create a TwoWaySwapDatum. Assets are sorted internally so order of
      * (first, second) doesn't matter — just match prices to the correct asset.
      */
    def genSwapDatum(
        firstId: ByteString, firstName: ByteString,
        secondId: ByteString, secondName: ByteString,
        firstPrice: Rational,
        secondPrice: Rational,
        prevInput: Option[TxOutRef] = Option.None,
        expiration: Option[BigInt] = Option.None
    ): TwoWaySwapDatum =
        // Sort assets; first < second means first becomes asset1
        val firstIsSmaller = compareAssets(firstId, firstName, secondId, secondName) <= 0
        val (a1Id, a1Name, a2Id, a2Name, a1Price, a2Price) =
            if firstIsSmaller
            then (firstId, firstName, secondId, secondName, firstPrice, secondPrice)
            else (secondId, secondName, firstId, firstName, secondPrice, firstPrice)

        TwoWaySwapDatum(
            beaconId = beaconCurrencySymbol,
            pairBeacon = genPairBeaconName(a1Id, a1Name, a2Id, a2Name),
            asset1Id = a1Id,
            asset1Name = a1Name,
            asset1Beacon = genAssetBeaconName(a1Id, a1Name),
            asset2Id = a2Id,
            asset2Name = a2Name,
            asset2Beacon = genAssetBeaconName(a2Id, a2Name),
            asset1Price = a1Price,
            asset2Price = a2Price,
            prevInput = prevInput,
            expiration = expiration
        )

    /** Determine the correct swap redeemer based on which asset you're offering/asking.
      * Matches Haskell getRequiredSwapDirection.
      */
    def getRequiredSwapDirection(
        offerId: ByteString, offerName: ByteString,
        askId: ByteString, askName: ByteString
    ): TwoWaySwapRedeemer =
        if compareAssets(offerId, offerName, askId, askName) <= 0
        then TwoWaySwapRedeemer.TakeAsset1
        else TwoWaySwapRedeemer.TakeAsset2
